proposals/1125_dynamic-supervisor-oidc-clients/README.md

```diff
@@ -0,0 +1,820 @@
+  on the link.
+
+#### Upgrades
+
+The proposed CRD will be new, so there aren't any upgrade concerns for it. Potential changes to the FederationDomain
+resource are also new additions to an existing resource, so there are again no upgrade concerns there.
+
+The `pinniped-cli` client needs to continue to act as before for backwards compatibility with existing installations of
+the Pinniped CLI on user's machines. Therefore, it should be excluded from any new scope-based requirements which would
+restrict the group memberships from being returned. This will allow mixing of old CLIs with new Supervisors, and new
+CLIs with old Supervisors, in regard to the new Supervisor features proposed herein.
+
+#### Tests
+
+As usual, unit tests should be added for all new/changed code.
+
+Integration tests should be added to mimic the usage patterns of a webapp. Dynamic clients should be configured with
+various options to ensure that the options work as expected. Those clients should be used to perform the authcode flow,
+RFC8693 token exchanges, and TokenCredentialRequest calls. Negative tests should include validation failures on the new
+CRs, and failures to perform actions that are supposed to be disallowed on the client by its configuration.
+
+#### New Dependencies
+
+None are foreseen.
+
+#### Performance Considerations
+
+Some considerations were mentioned previously for client secret option 2 above. No other performance impact is foreseen.
+
+#### Observability Considerations
+
+None are foreseen, aside from the usual error messages and log statements for the new features similar to what is
+already implemented in the Supervisor.
+
+#### Security Considerations
+
+Some security considerations were already mentioned above. Here are a couple more.
+
+##### Ensuring reasonable client secrets
+
+During a login flow, when the client secret is presented in plaintext to the Supervisor’s token endpoint, it could
+potentially validate that the secret meets some minimum entropy requirements. For example, it could check that the
+secret has sufficient length, a sufficient number of unique characters, and a sufficient number of letter vs number
+characters. If we choose to use the plaintext passwords option then the Supervisor could potentially perform this
+validation in the mutating admission webhook before it hashes the passwords.
+
+##### Disallowing audience confusion
+
+Who decides the names of the dynamic client and the workload clusters?
+
+- The name of the dynamic client would be chosen by the admin of the Supervisor. We could put validations on the name if
+  we would like to limit the allowed names.
+- The name of the workload cluster is chosen by the admin of the workload cluster (potentially a different person or
+  automated process). We don’t currently limit the string which chooses the audience name for the workload cluster, so
```

Comment:
Question/thought: why do clusters have an audience name at all?  Why is this not simply the URL of the cluster?

Reply:
Some related thoughts: a cluster could have multiple URLs, which is probably fine (sounds like something we can handle).  Uniqueness could be an issue, i.e. kind clusters can have the same URL, which would be a problem.  Maybe "audience" values that start with `https://` are special?